### PR TITLE
:bug: Automatically delete machines marked for deletion

### PR DESCRIPTION
**Pull request by wmgroot**
**What this PR does / why we need it**:
This PR is a sister PR to a change in Cluster Autoscaler. This is to handle an edge case where nodes that fail to join a cluster cause Cluster Autoscaler to downscale desired replica counts below what is necessary to schedule all pods in the cluster.

https://github.com/kubernetes/autoscaler/pull/5538

Please see the linked for detailed context on the Cluster Autoscaler changes.

TLDR: We want to support a method for Cluster Autoscaler to delete nodes that isn't based around downscaling desired replica counts, in order for it to cancel nodes that exceed its max-node-provision-time.

**Which issue(s) this PR fixes**:
Fixes this Cluster Autoscaler issue.
https://github.com/kubernetes/autoscaler/issues/5465

### CODE REVIEW

File: internal/controllers/machineset/machineset_controller.go [hunk 1]
```diff
@@ -306,16 +306,13 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
...
-			}
-			conditions.MarkTrue(machine, clusterv1.MachineOwnerRemediatedCondition)
-			if err := r.Client.Status().Patch(ctx, machine, patch); err != nil && !apierrors.IsNotFound(err) {
-				errs = append(errs, errors.Wrap(err, "failed to update status"))
-			}
+			log.Info("Deleting machine because marked as unhealthy by the MachineHealthCheck controller: %v", machine.Name)
+			errs = append(r.deleteMachine(ctx, machine))
+		}
+		// The delete-machines annotation may be set by external sources, such as Cluster Autoscaler
+		if _, ok := machine.ObjectMeta.Annotations[clusterv1.DeleteMachineAnnotation]; ok {
```
Comment (killianmuldoon) [inline internal/controllers/machineset/machineset_controller.go:313]: I think if this route is chosen there should be a new annotation created specifically for this purpose.. 

`clusterv1.DeleteMachineAnnotation` is already used below in syncReplicas in L567 to give priority to deletion. Using the same annotation here changes the behaviour in a way that's confusing.

Comment (wmgroot) [reply internal/controllers/machineset/machineset_controller.go:313]: Is the delete-machine annotation set by anything other than CA?

Comment (killianmuldoon) [reply internal/controllers/machineset/machineset_controller.go:313]: I don't know -  but it's a public annotation with a defined behaviour so I don't think we should change it here.

Comment (wmgroot) [reply internal/controllers/machineset/machineset_controller.go]: Sounds reasonable. I've added a new annotation `cluster.x-k8s.io/machine-marked-for-delete`.

**Comment by sbueringer** [comment]
Not sure if I understand correctly. Basically the autoscaler detects that node failed to join / didn't join quick enough and thus the autoscaler wants the machine/node to be recreated?

If this is what we want to do, the MachineHealthCheck controller is doing the same. If you don't want to use a MHC for it you might be able to signal the same way as the MHC (by setting a condition on the Machine)

**Comment by elmiko** [comment]
> Not sure if I understand correctly. Basically the autoscaler detects that a node failed to join / didn't join quick enough and thus the autoscaler wants the machine/node to be recreated?

sort of, the autoscaler generally has a single method for requesting the provider to delete a node. in cases where an instance has failed to become a node the autoscaler will want to tell the provider to remove that node, but in the case of cluster-api it also tries to reduce the number of replicas. this combination is causing a situation where far too many nodes get removed, and iiuc sometimes the wrong nodes get deleted.

this patch, in combination with https://github.com/kubernetes/autoscaler/pull/5538 is seeking to give us more control over how the machines should be deleted from the autoscaler side.

i'm hoping i got that correct, this is a complicated issue but it's something that is affecting @wmgroot . happy for any corrections.

**Comment by wmgroot** [comment]
The crux of the issue on the CAPI side is this.
* A Machine in Provisioning status exceeds CA's max-node-provision-time.
* CA annotates that Machine resource with the "delete-machine" annotation.
* CA drops the replica count of the MachineDeployment by 1.

At this time, if the MachineDeployment is migrating from one MachineSet to another, the MachineDeployment controller will choose to remove the extra Machine from the old MachineSet, rather than respecting the "delete-machine" annotation on the Provisioning status Machine in the new MachineSet.

If you are trying to cycle many Machines simultaneously, such as with a 100 node MachineDeployment + maxSurge=20% and the Machines all become stuck in Provisioning status (this can happen with AWS InsufficientCapacity issues, or simply slow capa-controller execution in our case), CA will aggressively drop the replica count of the MachineDeployment once for every Provisioning Machine (20 in this example). This behavior evicts running pods with no new nodes to schedule on, leaving you with a large capacity deficit in the cluster.

**Comment by sbueringer** [comment]
Apologies if I'm missing something in the linked PRs etc.

> sort of, the autoscaler generally has a single method for requesting the provider to delete a node. in cases where an instance has failed to become a node the autoscaler will want to tell the provider to remove that node, but in the case of cluster-api it also tries to reduce the number of replicas.

This sounds like it was a deliberate choice of the autoscaler Cluster API implementation to delete a node **and** reduce the replica number. What is the reason that it cannot just delete the Machine? (literally by just calling delete on the Machine it wants to delete?)

**Comment by elmiko** [comment]
i need to spend some more time reviewing the code in the autoscaler, i think @sbueringer makes a good comment about why we can't just delete the machine from the autoscaler side, but i need to review it again.

also, i think @wmgroot is making a very salient point as well:

> At this time, if the MachineDeployment is migrating from one MachineSet to another, the MachineDeployment controller will choose to remove the extra Machine from the old MachineSet, rather than respecting the "delete-machine" annotation on the Provisioning status Machine in the new MachineSet.

i don't know the mechanics behind the MachineSet migrations that occur, and how this interacts with the autoscaler requesting deletions. it seems like the deletion annotation should be respected in this condition?

**Comment by sbueringer** [comment]
> i don't know the mechanics behind the MachineSet migrations that occur, and how this interacts with the autoscaler requesting deletions. it seems like the deletion annotation should be respected in this condition?

When a MD is rolled out we scale up a new MS and scale down the old one. I didn't check it myself but based on the comments above when the autscaler downscales the MD the MD controller then downscales the old MachineSet This makes sense as the goal of the MD controller is to replace the old MachineSet with the new one and it should not look at individual Machines. 

This PR now essentially adds an annotation to a Machine and the MachineSet controller will delete the Machine with the next reconcile, 100% independent of any MD=>MS rollout logic.

**Comment by elmiko** [comment]
> This PR now essentially adds an annotation to a Machine and the MachineSet controller will delete the Machine with the next reconcile, 100% independent of any MD=>MS rollout logic.

should the MD controller be checking to see if there is a Machine with a deletion annotation to remove before scaling down the old MachineSet on a replica reduction?

> That's the part that I'm confused about. Instead of just deleting the Machine directly the autoscaler now tells the MachineSet controller to delete the Machine, which it will do instantly.

i need some time to investigate/think about this a little more. perhaps the autoscaler should be deleting machines in some cases, i need a little more context.

**Comment by sbueringer** [comment]
> should the MD controller be checking to see if there is a Machine with a deletion annotation to remove before scaling down the old MachineSet on a replica reduction?

We can open an issue to describe what exactly the expected behavior should be. I'm not sure if it's a good idea if we change the MD controller so it also acts on Machines and not just on MachineSets. Maybe there is something that should be done on the MachineSet level and then the MD controller can act on that when it sees the change in the MachineSet.

But I think what probably should not happen is that while the MD is in a rollout and when the MD replica count is reduced that we start scaling down the new MachineSet. After all our goal is to rollout from the old to the new MachineSet.

**Comment by wmgroot** [comment]
Yeah, I arrived at the same conclusion that downscaling the replica count of the MD in order to delete the node was not a desired behavior. On further thought, that's only true some of the time. It is a desired behavior if CA had previously incremented the replica count to scale up the node group, and is reducing the replica count in response to a node failing to join in order to try a different node group as a fallback.

In the second situation, the desired behavior is for CAPI to understand that the failing node is what should be deleted, rather than a node in the old MachineSet during a rollout.

This is pretty complex, and I'd be happy to schedule some time to talk through what we think the best options might be.

**Comment by sbueringer** [comment]
Let's try to get some input from other folks who are familiar with the current implementation given that it's pretty much the most complex part of our code base.

@vincepri @enxebre @fabriziopandini @ykakarap @killianmuldoon

**Comment by elmiko** [comment]
> We can open an issue to describe what exactly the expected behavior should be and discuss it in more detail. I'm not sure if it's a good idea if we change the MD controller so it also acts on Machines and not just on MachineSets. Maybe there is something that should be done on the MachineSet level and then the MD controller can act on that when it sees the change in the MachineSet.

that makes sense to me, i am not trying to propose that we add action which was never intended by the controller. but maybe there is some consideration that the MD controller needs to take when choosing the MS to scale down, and then the MS controller needs to prefer the annotated nodes or something.

**Comment by sbueringer** [comment]
The MD controller picks the MachineSet with the correct spec (so it can't just pick another one because it has more replicas).

If there are multiple MachineSets with a matching spec it would pick the one with more replicas. But that's more an edge case / migration case then the normal rollout behavior.

**Comment by elmiko** [comment]
> The MD controller picks the MachineSet with the correct spec (so it can't just pick another one because it has more replicas).

so if the user updates an MD, and thus causes a new rollout, while the autoscaler is also in the middle of a scale down operation, then the MD controller will only pick the new MS as it matches the spec? (i think this is the heart of @wmgroot 's issue)

**Comment by sbueringer** [comment]
> > The MD controller picks the MachineSet with the correct spec (so it can't just pick another one because it has more replicas).
> 
> so if the user updates an MD, and thus causes a new rollout, while the autoscaler is also in the middle of a scale down operation, then the MD controller will only pick the new MS as it matches the spec? (i think this is the heart of @wmgroot 's issue)

Yes it will only pick the new MS.
When a MD is changed in a way that in-place propagation is not enough to rollout the change (e.g. let's say someone changes the failureDomain) the MD controller creates a new MS and tries to scale the new MS up and the old MS down.

**Comment by wmgroot** [comment]
Here's the relevant code from the MD controller when a rollout is in progress.
https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machinedeployment/machinedeployment_rolling.go#L101-L181

The main things to note IMO are:
* `cleanupUnhealthyReplicas()` is only ever called for the old MS.
* `scaleDownOldMachineSetsForRollingUpdate()` can only scale down the old MS.
* The case 1 example accounts for new nodes failing to join, but assumes that `replicas` does not change.

In my team, we've talked about potentially having the MD controller be made smarter and count nodes marked for deletion as unhealthy to detect a scenario where the new MS should be scaled down instead of the old MS, but I'd like to hear others' thoughts too.

**Comment by sbueringer** [comment]
> On further thought, that's only true some of the time. It is a desired behavior if CA had previously incremented the replica count to scale up the node group, and is reducing the replica count in response to a node failing to join in order to try a different node group as a fallback.

> In the second situation, the desired behavior is for CAPI to understand that the failing node is what should be deleted, rather than a node in the old MachineSet during a rollout.

Q: Do I interpret that correctly as it means that the CA intentionally scales down the MD with the intention of scaling down the new MachineSet? How does the CA try to achieve this: "try a different node group as a fallback"?

**Comment by elmiko** [comment]
> Q: Do I interpret that correctly as it means that the CA intentionally scales down the MD with the intention of scaling down the new MachineSet and scaling back up the old MS? How does the CA try to achieve this: "try a different node group as a fallback"? It doesn't actualy do anything after scale down which goes in the direction of trying to rollback to the old MachineSet right?

from the autoscaler's perspective, it just sees "node groups", which in cluster-api world maps directly to either a MachineDeployment or a MachineSet. the core autoscaler then performs actions that are like "increase the size of this node group by X" , "decrease the size of this node group by Y" , or "delete nodes A, B, and C". for our provider we mostly only change the replica counts in response to scaling events, with the exception of the delete nodes interface. for that we place annotations on the machines we want deleted and let the cluster api controllers do the rest of the work.

essentially we have tried to make the autoscaler implementation be very basic about what it does to capi objects, we have purposely left the decisions for how machines are created/deleted to the capi controllers.

now, given that, we might be able to add more intelligence to the autoscaler side of things, but i would want to make sure that we are bubbling internal details from capi into the autoscaler as it becomes more tightly coupled.

**Comment by sbueringer** [comment]
Just to clarify

> The crux of the issue on the CAPI side is this.
> * A Machine in Provisioning status exceeds CA's max-node-provision-time.
> * CA annotates that Machine resource with the "delete-machine" annotation.
> * CA drops the replica count of the MachineDeployment by 1.

The last step in this sequence is not to get rid of this Machine it's just something independent that can also happen?

But that means in the normal case after step 1+2 the Machine is not deleted at all, right?

**Comment by wmgroot** [comment]
> The last step in this sequence is not to get rid of this Machine it's just something independent that can also happen?
It drops the replica count in order to delete that specific Machine, knowing that CAPI will prefer to delete the annotated Machine.

Here's the actual code for the CAPI implementation of the DeleteNodes() function.
https://github.com/kubernetes/autoscaler/blob/b1b7f3990893947692a978f9b36145f5e6fe0c59/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go#L132-L166

My patched version of the interaction prevents the replica count from being dropped since CA can assume that CAPI will clean up the annotated Machine instead of needing to drop the replica count to do it.

**Comment by enxebre** [comment]
@wmgroot, @sbueringer, @elmiko  Let's try to break this down into specific problems and potential solutions.

1 - During a MachineDeployment rollout, new machines might find a problem that prevent them from coming up. In that scenario if a consumer scales down the MachineDeployment, it will prefer to delete Machines from the old MachineSet, therefore causing the loss of healthy Machines and increasing unavailability. Did I get it right?

2 - Now, a consumer (e.g. the autoscaler) might think it can't trust the delete-machine annotation, but this is not the case during the middle of a MachineDeployment rolling upgrade (as described above).

>Cluster Autoscaler to delete nodes that isn't based around downscaling desired replica counts, in order for it to cancel nodes that exceed its max-node-provision-time.

3- If this ^^ is a problem beyond 2, let's call it 3 and this needs to be discussed generically (not only for capi provider) in autoscaler. The expectation at the moment is for DeleteNodes() to decrease replica of the NodeGroup.
See https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L179-L182
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go#L321

**Comment by elmiko** [comment]
i generally agree with the way you've broken it down @enxebre . imo, i would expect decreasing the replica count to take into account the delete-machine annotation, but if there are cases where it is specifically /not/ observed then the autoscaler (or any consumer) would need some way to know when that would happen. at least, that's what i'm thinking currently.

**Comment by fabriziopandini** [comment]
thanks @elmiko for the wrap-up
What about moving point 1 into an issue so we can discuss options?

**Comment by vincepri** [comment]
Closing this due to inactivity, feel free to reopen when ready to discuss again.

/close
